### PR TITLE
[plugin-web-app] Add keys combination pressing functionality

### LINKED ISSUE #686: Unable to delete text through by ctrl+a+delete
Unable to delete text from field using step "When I execute sequence of actions", steps are passed, but instead of deleting text - character "A" is added to the field. 

Example story:
[test.txt](https://github.com/vividus-framework/vividus/files/4805424/test.txt)

**Expected behavior**
After execution sequence of actions the whole text in context area is deleted

**Artifacts**
input_a
[allure.zip](https://github.com/vividus-framework/vividus/files/4805453/allure.zip)

**Execution environment (please complete the following information):**
 - Vividus version: 0.2.4-SNAPSHOT
 - OS: macOS Mojave 10.14.6
 - Java: java 13.0.2

**Additional context**
Priority - p2

### PR DESCRIPTION
**Pull request by VitaliLebedzeu**
Fixes #686

### CODE REVIEW

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java [hunk 1]
```diff
@@ -129,6 +132,27 @@ public Type getArgumentType()
...
+        private final Type argumentType = TypeUtils.parameterize(List.class, String.class);
+
+        @Override
+        public void addAction(Actions actions, Object argument)
+        {
+            perform(argument, (List<String> arg) -> {
+                LinkedList<String> linkedArg = new LinkedList<>(arg);
+                String firstKey = linkedArg.pollFirst();
+                Keys pressedKey = EnumUtils.isValidEnum(Keys.class, firstKey) ? Keys.valueOf(firstKey) : null;
+                actions.keyDown(pressedKey).sendKeys(KeysUtils.keysToCharSequenceArray(linkedArg)).keyUp(pressedKey);
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:146]: passing null as pressedKey will lead to IllegalArgumentException or NPE as per Actions logic

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java [hunk 2]
```diff
@@ -129,6 +132,27 @@ public Type getArgumentType()
...
     },
+    PRESS_KEYS_SIMULTANEOUSLY(false)
+    {
+        private final Type argumentType = TypeUtils.parameterize(List.class, String.class);
+
+        @Override
+        public void addAction(Actions actions, Object argument)
+        {
+            perform(argument, (List<String> arg) -> {
+                LinkedList<String> linkedArg = new LinkedList<>(arg);
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:143]: why do you need LinkedList here? if list is empty, null returned by pollFirst will not be suitable for keyDown

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java [hunk 3]
```diff
@@ -175,6 +176,8 @@ public void executeActionsSequence(List<Action> actions)
      * <td>PRESS_KEYS</td>
      * <td><a href="https://selenium.dev/selenium/docs/api/java/org/openqa/selenium/Keys.html">Keys</a>
      * </td>
+     * <td>PRESS_KEYS_SIMULTANEOUSLY</td>
+     * <td>press keys combination such as "ctrl + a" etc.</td>
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java:180]: can I press more than two keys?

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/model/SequenceActionTypeTests.java [hunk 4]
```diff
@@ -130,6 +132,19 @@ void testPressKeys()
         verifyNoMoreInteractions(baseAction);
     }
 
+    @Test
+    void testPressKeysSimultaneously()
+    {
+        Mockito.when(baseAction.keyDown(Mockito.any(Keys.class))).thenReturn(baseAction);
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/model/SequenceActionTypeTests.java:138]: use static imports for mockito methods

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java [hunk 5]
```diff
@@ -116,19 +116,20 @@ public void executeActionsSequence(List<Action> actions)
...
+     * <br> |CLICK_AND_HOLD           |By.xpath(//signature-pad-control/canvas)|
+     * <br> |CLICK_AND_HOLD           |                                        |
+     * <br> |MOVE_BY_OFFSET           |(-300, 0)                               |
+     * <br> |RELEASE                  |By.xpath(//signature-pad-control/canvas)|
+     * <br> |RELEASE                  |                                        |
+     * <br> |ENTER_TEXT               |Text                                    |
+     * <br> |CLICK                    |By.placeholder(Enter your password)     |
+     * <br> |CLICK                    |                                        |
+     * <br> |PRESS_KEYS               |BACK_SPACE                              |
+     * <br> |PRESS_KEYS_SIMULTANEOUSLY|CONTROL,a                               |
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java:131]: unit test for ActionSteps should be added as well

**Review by uarlouski — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java (hunk 1 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:146]: Yes. I try to keep the logic that wrong keys order should abort the method and throws the runtime exception

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java (hunk 2 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:143]: I want to use the "LinkedList" for managing the base list and getting and removing the first element simultaneously. If the list will be empty or null I expect the same result as was mentioned in previous [https://github.com/vividus-framework/vividus/pull/833#discussion_r472916456](comment)

Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:143]: But considering the "DeepCode" issue, I think I will continue to find another solution.

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java (hunk 3 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java:180]: If you mean pressing more than one key simultaneously (Example "ctrl + alt + delete") I think no, you can't. For this issue, I've seen another solution than every key that you want to press simultaneously should be realized by the "keyDown()" method.

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/model/SequenceActionTypeTests.java (hunk 4 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/model/SequenceActionTypeTests.java:138]: fixed

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java (hunk 1 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java]: refactored

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java (hunk 2 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java]: refactored

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java (hunk 3 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java]: rewrite

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java [hunk 6]
```diff
@@ -129,6 +132,36 @@ public Type getArgumentType()
...
     },
+    PRESS_KEYS_SIMULTANEOUSLY(false)
+    {
+        private final Type argumentType = TypeUtils.parameterize(List.class, String.class);
+
+        @Override
+        public void addAction(Actions actions, Object argument)
+        {
+            perform(argument, (List<String> arg) -> {
+                List<String> arrayArg = new ArrayList<>(arg);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:147]: `arrayArg` -> `args`

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java [hunk 7]
```diff
@@ -129,6 +132,36 @@ public Type getArgumentType()
...
+                String firstKey = arrayArg.get(0);
+                arrayArg.remove(0);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:149]:
```
String firstKey = arrayArg.remove(0);
```

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java [hunk 8]
```diff
@@ -116,19 +116,20 @@ public void executeActionsSequence(List<Action> actions)
...
+     * <br> |MOVE_TO                  |By.id(name)                             |
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java:132]: Would you mind adding the documentation?
see the guidelines: https://github.com/vividus-framework/docs.vividus.dev#one-liner

Comment (valfirst) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java:132]: also you can find parts of the docs in the release notes

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java (hunk 6 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:147]: fixed

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java (hunk 7 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:149]: fixed

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java [hunk 9]
```diff
@@ -129,6 +132,40 @@ public Type getArgumentType()
...
     },
+    PRESS_KEYS_SIMULTANEOUSLY(false)
+    {
+        private final Type argumentType = TypeUtils.parameterize(List.class, String.class);
+
+        @Override
+        public void addAction(Actions actions, Object argument)
+        {
+            perform(argument, (List<String> arg) -> {
+                if (arg.size() == 0)
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:143]: `org.apache.commons.lang3.Validate#notEmpty(T, java.lang.String, java.lang.Object...)`

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java [hunk 10]
```diff
@@ -129,6 +132,40 @@ public Type getArgumentType()
...
+                {
+                    throw new IllegalArgumentException("At least one key should be provided. "
+                            + "The provided keys should start with of 'SHIFT', 'ALT' or 'CONTROL'/'COMMAND' key.");
+                }
+                List<String> args = new ArrayList<>(arg);
+                String firstKey = args.remove(0);
+                if (EnumUtils.isValidEnum(Keys.class, firstKey))
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:150]: `org.apache.commons.lang3.Validate#isTrue(boolean, java.lang.String, java.lang.Object...)`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java (hunk 9 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java:143]: fixed

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java (hunk 10 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/model/SequenceActionType.java]: fixed

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java (hunk 8 above)
Comment (VitaliLebedzeu) [reply vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/ActionSteps.java:132]: added

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 11]
```diff
@@ -87,6 +87,40 @@ Then number of HTTP GET requests with URL pattern `http://httpbin.org/get` is EQ
...
+==== *_Wording_*
+
+[source,gherkin]
+----
+When I execute sequence of actions: $actions
+----
+
+==== *_Parameters_*
+
+. `$actions` - table of actions to execute
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:105]: could you please describe the actions? see javadocs or release notes for details

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/plugins/pages/plugin-web-app.adoc (hunk 11 above)
Comment (VitaliLebedzeu) [reply docs/modules/plugins/pages/plugin-web-app.adoc]: described

**Review by valfirst — APPROVED** [review]

**Pull request closed without merging**